Develop

MChepulis opened the pull request:
Обновил README.md файл
добавил описание, постановку задачи, требования к изображения
добавил отчёт по первой итерации (Big_Lab_Repport.ipynb)
добавил код первой итерации (main.py)

MChepulis committed:
add dataset

MChepulis committed:
add Task and Plan

MChepulis committed:
add First iteration

add report (.ipynb)
add python code

MChepulis committed:
add first iteration

change README file
  add project dicription
  add task
  add setup instruction

moved main.py
fix remark in Task.txt

Nikitagritsaenko reviewed (commented):
Вы сильно усложняете себе задачу таким датасетом. У Вас есть фотографии с шумом/без шума, засветлённые/затемнённые/с нормальным освещением, с сильной перспективой двери/без перспективы, с разных точек съёмки и пр. 

Как я понимаю, Вы пока пробовали находить стул только для контрастной фотографии с хорошим освещением. Но что будет в других случаях? Скорее всего, чтобы подсчитать accuracy Вашего итогового алгоритма, Вам придётся сделать новый датасет с более "пригодными" фотографиями. Потому что в противном случае будет трудно разработать алгоритм, работающий лучше, чем "монетка".

Nikitagritsaenko reviewed (commented):
У Вас дважды добавлены файлы Task.txt и Plan.txt. Один раз в коммите "add Task and Plan", а во второй раз в коммите с первой итерацией. При этом файлы ничем не отличаются друг от друга. Нужно удалить копии, иначе возникает путаница.

Кроме того, в этих файлах очень много опечаток.

Nikitagritsaenko commented on Task.txt:
```diff
@@ -0,0 +1,19 @@
+Выход: "да"/"нет"/"не знаю"
+	"да"  - в данной ориентации стул А проходит в дверной проём Б
+	"нет" - в данной ориентации стул А не проходит в дверной проём Б
+	"да"  - не удалось выбрать между "да"/"нет", изображение не удовлетворяет требованиям
```
"не знаю"*

MChepulis replied:
согласен, вкралась опечатка

Nikitagritsaenko commented on Plan.txt:
```diff
@@ -0,0 +1,22 @@
+    1) не зашумлены лишними предметами
+    2) имеют "удобное" освещение (оценивается разработчиком)
+    3) на фото обязательно присутствуют и стул и дверь
+- Ответ будет только "да" / "нет" (исключается ответ "не знаю")
```
> Ответ будет только "да" / "нет" (исключается ответ "не знаю")

В README написано иначе: может быть 3 ответа. Так как Вы всё-таки решили?

MChepulis replied:
в файле Plan.txt находится план на первую итерацию
В этой итерации не рассматриваются многие аспекты. В частности не обрабатываются изображения на которые можно было бы ответить "не знаю". Потому множество возможных  ответов сократилось до {"да", "нет"}

Nikitagritsaenko reviewed (commented):
При работе с путями используйте метод os.path.join. Он удобный и платформо-безопасный.

Например, в [3] кодовой секции, Вы пишите:

```python
name = "yes_ok_4.jpg"
data_img = imread("%s\\%s" % (dataset_path, name))
```

Можно написать вот так:

```python
name = "yes_ok_4.jpg"
path = os.path.join(dataset_path, name)
data_img = imread(path)
```

Nikitagritsaenko commented:
В [9] секции Вы применяете преобразование Хафа для прямых, чтобы найти границы двери. Но у Вас также находятся и горизонтальные прямые. Это помешает Вам найти ширину дверного проёма. Однако Вы легко можете проверить каждую прямую на угол наклона.

Например, вот так:

```python
vertical_limiter = 0.3 # позволяет среди всех прямых найти вертикальные
if (np.abs(angle) <= vertical_limiter):
    # только такие прямые рисовать и использовать для поиска ширины двери

```

UPD:
Как я понял, Вы находите дверь ORB детектором, избавляясь при этом от перспективы. Нужно тогда написать, какой метод Вы в итоге выбрали для своего алгоритма - Хаф или ORB.

Но стоит заметить, что warpPerspective для двери у Вас не всегда работает корректно. Скорее всего Хаф для двери отработает лучше.

MChepulis commented:
> У Вас дважды добавлены файлы Task.txt и Plan.txt. Один раз в коммите "add Task and Plan", а во второй раз в коммите с первой итерацией. При этом файлы ничем не отличаются друг от друга. Нужно удалить копии, иначе возникает путаница.
> 
> Кроме того, в этих файлах очень много опечаток.

По поводу дублирования файлов: Раньше в них был глубокий сакральный смысл, который заключался в том, чтобы папка с итерацией содержала бы задание и план конкретно на эту итерацию. Но сейчас осознал, что действительно, на каждую итерацию не стоит заводить отдельную папку.

Лишнее удалю, опечатки исправлю

MChepulis commented:
> Вы сильно усложняете себе задачу таким датасетом. У Вас есть фотографии с шумом/без шума, засветлённые/затемнённые/с нормальным освещением, с сильной перспективой двери/без перспективы, с разных точек съёмки и пр.
> 
> Как я понимаю, Вы пока пробовали находить стул только для контрастной фотографии с хорошим освещением. Но что будет в других случаях? Скорее всего, чтобы подсчитать accuracy Вашего итогового алгоритма, Вам придётся сделать новый датасет с более "пригодными" фотографиями. Потому что в противном случае будет трудно разработать алгоритм, работающий лучше, чем "монетка".

MChepulis commented:
> При работе с путями используйте метод os.path.join. Он удобный и платформо-безопасный.
> 
> Например, в [3] кодовой секции, Вы пишите:
> 
> ```python
> name = "yes_ok_4.jpg"
> data_img = imread("%s\\%s" % (dataset_path, name))
> ```
> 
> Можно написать вот так:
> 
> ```python
> name = "yes_ok_4.jpg"
> path = os.path.join(dataset_path, name)
> data_img = imread(path)
> ```

спасибо, учту это в следующей итерации

MChepulis commented:
> Вы сильно усложняете себе задачу таким датасетом. У Вас есть фотографии с шумом/без шума, засветлённые/затемнённые/с нормальным освещением, с сильной перспективой двери/без перспективы, с разных точек съёмки и пр.
> 
> Как я понимаю, Вы пока пробовали находить стул только для контрастной фотографии с хорошим освещением. Но что будет в других случаях? Скорее всего, чтобы подсчитать accuracy Вашего итогового алгоритма, Вам придётся сделать новый датасет с более "пригодными" фотографиями. Потому что в противном случае будет трудно разработать алгоритм, работающий лучше, чем "монетка".

В папке dataset лежат все фотографии, собранные для этого задания.
В следующей итерации планирую сделать так, чтобы алгоритм получал на вход некоторый путь до папки с данными и вызывался для всех изображений в этой папке.
Так будет легче контролировать работу на определённых изображения (случаях)

MChepulis commented:
> В [9] секции Вы применяете преобразование Хафа для прямых, чтобы найти границы двери. Но у Вас также находятся и горизонтальные прямые. Это помешает Вам найти ширину дверного проёма. Однако Вы легко можете проверить каждую прямую на угол наклона.
> 
> Например, вот так:
> 
> ```python
> vertical_limiter = 0.3 # позволяет среди всех прямых найти вертикальные
> if (np.abs(angle) <= vertical_limiter):
>     # только такие прямые рисовать и использовать для поиска ширины двери
> ```
> 
> UPD:
> Как я понял, Вы находите дверь ORB детектором, избавляясь при этом от перспективы. Нужно тогда написать, какой метод Вы в итоге выбрали для своего алгоритма - Хаф или ORB.
> 
> Но стоит заметить, что warpPerspective для двери у Вас не всегда работает корректно. Скорее всего Хаф для двери отработает лучше.

Да, искать дверь при помощи ORB детектора оказалось весьма проблематично. Особенно, учитывая то, что часть двери обычно перекрыта стулом
Сначала я попробовал использовать алгоритм Хафа, но не смог придумать, как среди полученных линий выбрать нужные.
Безбоязненно можно отбросить "точно не вертикальные"
Основную проблему вызывает дверной косяк и линии, полученные от ножек стула

Nikitagritsaenko commented:
> Основную проблему вызывает дверной косяк и линии, полученные от ножек стула

Не думаю, что это будет для Вас большой проблемой. Вы можете:

а) Применить вероятностное преобразование Хафа, в котором позволяется указывать минимальную и максимальную длину найденной линии. Таким образом, зная, что у двери линии длинные, а у стула - короткие, Вы сможете отсеять линии ножек стула.

б) Применить обычный поиск прямых Хафом на верхней части картинки, где стула и вовсе нет. Это кажется мне более простым решением.

Плинтус двери учесть легко, поскольку он занимает по ширине двери какую-то постоянную величину (скажем, 10 процентов). Причём эта величина примерно одинаковая для любой двери. Просто попробуйте учесть эту поправку.

MChepulis commented:
> > Основную проблему вызывает дверной косяк и линии, полученные от ножек стула
> 
> Не думаю, что это будет для Вас большой проблемой. Вы можете:
> 
> а) Применить вероятностное преобразование Хафа, в котором позволяется указывать минимальную и максимальную длину найденной линии. Таким образом, зная, что у двери линии длинные, а у стула - короткие, Вы сможете отсеять линии ножек стула.
> 
> б) Применить обычный поиск прямых Хафом на верхней части картинки, где стула и вовсе нет. Это кажется мне более простым решением.
> 
> Плинтус двери учесть легко, поскольку он занимает по ширине двери какую-то постоянную величину (скажем, 10 процентов). Причём эта величина примерно одинаковая для любой двери. Просто попробуйте учесть эту поправку.

Действительно, идея искать дверь там, где точно нет стула, звучит многообещающе.
Возможно, если искать сперва стул, а только потом уже дверь, то можно стул про"вырезать" с изображения
А по поводу плинтусов:  из-за того, что дверь  одного цвета с плинтусом, то не всегда понятно, линия, которую нашёл алгоритм относится к двери (внутренняя граница плинтуса) или к плинтусу (внешняя границ плинтуса)

Nikitagritsaenko commented:
> А по поводу плинтусов: из-за того, что дверь одного цвета с плинтусом, то не всегда понятно, линия, которую нашёл алгоритм относится к двери (внутренняя граница плинтуса) или к плинтусу (внешняя границ плинтуса)

По крайней мере в Вашей реализации внешние границы плинтуса находятся всегда, так что можете просто считать расстояние между самой левой и правой линией, а затем учитывать поправку на плинтус.

MChepulis commented:
> > > Основную проблему вызывает дверной косяк и линии, полученные от ножек стула
> > 
> > 
> > Не думаю, что это будет для Вас большой проблемой. Вы можете:
> > а) Применить вероятностное преобразование Хафа, в котором позволяется указывать минимальную и максимальную длину найденной линии. Таким образом, зная, что у двери линии длинные, а у стула - короткие, Вы сможете отсеять линии ножек стула.
> > б) Применить обычный поиск прямых Хафом на верхней части картинки, где стула и вовсе нет. Это кажется мне более простым решением.
> > Плинтус двери учесть легко, поскольку он занимает по ширине двери какую-то постоянную величину (скажем, 10 процентов). Причём эта величина примерно одинаковая для любой двери. Просто попробуйте учесть эту поправку.
> 
> Действительно, идея искать дверь там, где точно нет стула, звучит многообещающе.
> Возможно, если искать сперва стул, а только потом уже дверь, то можно стул про"вырезать" с изображения
> А по поводу плинтусов: из-за того, что дверь одного цвета с плинтусом, то не всегда понятно, линия, которую нашёл алгоритм относится к двери (внутренняя граница плинтуса) или к плинтусу (внешняя границ плинтуса)

Спасибо. Учитывая то, что на стене, действительно, нет других больших вертикальных линий это должно сработать.
Теперь нужно не забыть указать данное свойство, как требование к фотографии)

MChepulis commented:
> > А по поводу плинтусов: из-за того, что дверь одного цвета с плинтусом, то не всегда понятно, линия, которую нашёл алгоритм относится к двери (внутренняя граница плинтуса) или к плинтусу (внешняя границ плинтуса)
> 
> По крайней мере в Вашей реализации внешние границы плинтуса находятся всегда, так что можете просто считать расстояние между самой левой и правой линией, а затем учитывать поправку на плинтус.

Спасибо. Учитывая то, что на стене, действительно, нет других больших вертикальных линий это должно сработать.
Теперь нужно не забыть указать данное свойство, как требование к фотографии)

dimaaa1fed commented on Plan.txt:
```diff
@@ -0,0 +1,22 @@
+План
```
Согласен с Никитой по поводу датасета.
Сейчас датасет загромождён лишними фотографиями, такими как фотографиями табуретки, а не стула, который в итоге ищется, и совсем какими-то крайними случаями (например фотографии в темноте / открытой дверью). На всех этих фотографиях алгоритм ведет себя мало адекватно. Думаю, что стоит переработать датасет, предоставив больше требований к изображению.

MChepulis replied:
Согласен,
во первых программу планируется перевести на работу с конкретной папкой
во вторых, надо будет и избавиться от части фотографий, и, может быть, добавить новые
Также, будут переработаны требования к изображению

dimaaa1fed commented on First_iter/First_iter/main.py:
```diff
@@ -0,0 +1,244 @@
+
+# TODO скорее всего, матрицу обратного преобразования мужно найти при помощи матричных операций
+# TODO скорее всего, вместо преобразования маски можно преобразовывать прямогольники, однако, если маска будет более
+#  сложной формы, то прямоугольником уже не обойтись
```
Поиск стула сейчас осуществляется по особым точкам. На таких фотографиях, где стул повернут, также выдаются плохие результаты. (ex no_ok_1). Это происходит, так как похожих особых точек на таких фотографиях мало, они непонятно как расположены относительно друг друга, а следовательно и гомография будет находится кривая. Думаю стоит в обучающую выборку добавить много фотографий с различными ориентациями стула, и научить алгоритм искать изображение с похожей ориентацией (с помощью особых точек это возможно сделать).

MChepulis replied:
спасибо, попробую

dimaaa1fed commented on First_iter/First_iter/main.py:
```diff
@@ -0,0 +1,244 @@
+from matplotlib.patches import Rectangle
+
+
+dataset_path = "..\\dataset"
```
В .ipynb файле на этом месте неправильный путь (dataset_path = "..\Big_Lab\dataset")

MChepulis replied:
спасибо, исправлю

MChepulis committed:
add second iteration

add lab report in README
add images for lab report
add arguments
add requirements
add chair templates
add chair_finder_ORB
add chair_finder_bin
add door_finder
